### PR TITLE
[BACK-907] Add graphQL API -> resolver integration tests (CollectionAuthor)

### PR DESCRIPTION
**Pull request by nina-py**
## Goal

Add more integration tests to make sure no GraphQL query or mutation is broken by future development. I would love feedback on everything from where these tests should live and how they should be structured to any edge cases I have forgotten about. 

This is a pilot PR that sets everything up - no new packages but there's a new job in CircleCI, `test_server_integrations`. Once merged, we should make it required in the branch protection settings on `main` (@kkyeboah and I did this already during today's pairing but that made all outstanding and even merged PRs go yellow dot as they were waiting for the outcome of this job that's still on a branch, so I removed it for now).

## Todos

- [x] Add tests for `getCollectionAuthors`, `getCollectionAuthor` queries.
- [x] Add tests for `createCollectionAuthor` mutation.
- [x] Add tests for `updateCollectionAuthor` and `updateCollectionAuthorImageUrl` mutations.
- [x] Current concerns:
- ~Jest not exiting in CircleCI on completion of tests (fixed locally but persists on CircleCI) - temp fix implemented with `--forceExit` option, should look for a permanent fix in a follow-up PR? Details of the errors can be found [here](https://app.circleci.com/pipelines/github/Pocket/collection-api/1590/workflows/9afca870-725c-4ba8-8a8b-ff518ac4c429/jobs/8227/parallel-runs/0/steps/0-106).~ **Fixed by removing the Sentry plugin from test server config + starting and stopping Apollo Server for every test so that Apollo mops up hanging connections instead of Jest via --forceExit).**
- Unwanted output to console by Apollo during tests - it's a feature to make these errors easier to capture by CloudWatch, **fine to stay**.
- Complete suite of tests potentially taking 2 x forever rather than the 1 x forever integration tests take right now - split into two jobs? **Done**

Tickets:

- https://getpocket.atlassian.net/browse/BACK-907

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,7 +8,8 @@
     "start:dev": "npm run migrate:deploy && npm run build && npm run watch",
     "test-ci": "npm test",
     "test": "jest \"\\.spec\\.ts\"",
-    "test-integrations": "jest \"\\.integration\\.ts\" --runInBand",
+    "test-integrations": "jest --testPathPattern=src/database --runInBand",
```
[the remark this answers is not in the document]
Comment (nina-py) [reply package.json:11]: Well, the reason for splitting them up was to speed up the checks on PRs - since these tests use the database, too, they are slower than pure unit tests. I could tweak the path pattern so that it doesn't take in that extra test.

File: src/test/admin-server/queries/CollectionAuthor.server.integration.ts [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+          totalPages
+          totalResults
+          perPage
+        }
+      }
+    }
+  `;
+
+  beforeAll(async () => {
+    await clearDb(db);
```
[the remark this answers is not in the document]
Comment (nina-py) [reply src/test/admin-server/queries/CollectionAuthor.server.integration.ts:33]: Yes, I think that for testing queries it's not necessary to go back to zero in-between tests, I'm only querying data here and not manipulating it in any way in each of the tests.

**Comment by nina-py** [comment]
I agree that these tests repeat a lot of the assertions of the Prisma integration tests, but I feel that we need that there for completion. The bugs we've had previously were mainly with queries or mutations not returning some of the data - i.e. not returning story authors after a mutation on a story, I wonder if we could limit the scope of these tests to that instead of meticulously testing all the scalar fields as well? 

Also, we can do mocks ( for example, using this approach: https://dev.to/ekafyi/typing-and-mocking-a-graphql-api-server-with-apollo-3f4h ) and that would speed up the tests themselves, there would be no need to split them up from the database integration tests into their own job but... we already have the helpers for inserting data into the database, mocking has its advantages but it's almost easier to reuse what's already been set up for resolver testing.

File: src/test/admin-server/mutations/CollectionAuthor.server.integration.ts [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+      });
+
+      // Attempt to use the same slug one more time...
+      await expect(
+        server.executeOperation({
+          query: CREATE_COLLECTION_AUTHOR,
+          variables,
+        })
+        // ...without success
+      ).rejects.toThrowError();
```
[the remark this answers is not in the document]
Comment (nina-py) [reply src/test/admin-server/mutations/CollectionAuthor.server.integration.ts:108]: There is something there for sure. Context? We don't seem to be setting any beyond the db client and S3. Could S3 be the culprit on CircleCI? The tests exit cleanly in local development. Tried disabling the playground. 

One last thing left to try would be set up a separate server config for these tests instead of reusing the existing one, and strip absolutely everything out of there except the db client. If we decide to mock data as well, we could get rid of the db client as well there.

[the remark this answers is not in the document]
Comment (nina-py) [reply src/test/admin-server/mutations/CollectionAuthor.server.integration.ts]: I got to the bottom of the wrong error issue - once I removed the Sentry plugin from the server config used in these tests, I started getting the expected errors - either custom ones emitted by the resolvers like the one about duplicate slugs, or standard GraphQL ones for missing variables in inputs, etc. I've updated the tests to expect the right things to happen. 

However (!), the tests still hang on CircleCI.

[the remark this answers is not in the document]
Comment (nina-py) [reply src/test/admin-server/mutations/CollectionAuthor.server.integration.ts]: Updated the PR description - now stopping & starting Apollo Server after every test - this mops up open connections.

File: src/test/admin-server/queries/CollectionAuthor.server.integration.ts [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+          }
+          pagination {
+            currentPage
+            totalPages
+            totalResults
+            perPage
+          }
+        }
+      }
+    `;
```
Comment (jpetto) [inline src/test/admin-server/queries/CollectionAuthor.server.integration.ts:41]: random thought maybe for a fun friday - what if we had these sample queries in their own file for easy reference for use both here and manually in the playground?

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (jpetto) [reply package.json:11]: since the new server integration tests are already in a designated folder, what if we switch this logic around so `utils.spec.ts` (and any other future non-integration tests that end up `src/database`) don't get run twice? something like:

```
"test-integrations": "jest \"\\.integration\\.ts\" --runInBand",
"test-server-integrations": "jest --testPathPattern=src/test/admin-server --runInBand",
```

**Review by jpetto — CHANGES_REQUESTED** [review]
a couple very minor requests, but i think we're about ready to go here. i'm still on the fence around the duplicated tests, but we can iron that out in follow-up prs.

File: src/admin/server.ts [hunk 5]
```diff
@@ -6,6 +6,9 @@ import { sentryPlugin } from '@pocket-tools/apollo-utils';
 import { client } from '../database/client';
 import s3 from '../aws/s3';
 
+// Export this separately so that it can be used in Apollo integration tests
+export const db = client();
```
[the remark this answers is not in the document]
Comment (nina-py) [reply src/admin/server.ts:10]: Not any more, thank you for picking this up!

File: src/test/admin-server/queries/CollectionAuthor.server.integration.ts (hunk 4 above)
Comment (nina-py) [reply src/test/admin-server/queries/CollectionAuthor.server.integration.ts:41]: Good idea! Moved them out to their own files (one for queries, one for mutations). This made the test suites considerably shorter and easier to navigate :).

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (nina-py) [reply package.json]: Well, the new files (the ones added in this PR) still have to be either `.integration.ts` or `.spec.ts` so they would be picked up by the `test-integrations` commands. OR we could update the Jest config file which I'm reluctant to do - I'd rather it is in sync with the backend template.

After experimenting with patterns (and discovering there are integration tests in `src/aws`, too), I decided to ditch the second test run and just add the new tests to the old integration tests run. The wait time for will not expand by much - it is, despite my initial concern, not as important as I thought. 

I also looked up ways of suppressing console output in the new tests - there are lots of options, from running Jest with the `--silent` option to mocking parts of the global `console` object, but left everything as is for now. https://stackoverflow.com/questions/44467657/jest-better-way-to-disable-console-inside-unit-tests

**Review by jpetto — APPROVED** [review]
this looks great! thanks for seeing this through 😄
